src/ts/columnController.ts

```diff
@@ -620,41 +622,135 @@ module ag.grid {
             }
         }
 
-        private buildGroups() {
+        private isGroupVisible(columnGroup: ColumnGroup): boolean {
+            for (var i = 0; i < columnGroup.allColumns.length; i++) {
+                var column = columnGroup.allColumns[i];
+                if (this.visibleColumns.indexOf(column) > -1) {
+                    return true;
+                }
+            }
+            for (var j = 0; j < columnGroup.allSubGroups.length; j++) {
+                var subGroup = columnGroup.allSubGroups[j];
+                if (this.isGroupVisible(subGroup)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private splitColumnGroupForPinning(columnGroup: ColumnGroup, pinnedCols: Column[], unpinnedCols: Column[]): ColumnGroup[] {
+            var pinnedGroup = new ColumnGroup(true, columnGroup.name);
+            pinnedCols.forEach(function(col) {
+                pinnedGroup.addColumn(col);
+            });
+            var unpinnedGroup = new ColumnGroup(false, columnGroup.name);
+            unpinnedCols.forEach(function(col) {
+                unpinnedGroup.addColumn(col);
+            });
+            return [pinnedGroup, unpinnedGroup];
+        }
+
+        private checkForPinningInColumnGroup(columnGroup: ColumnGroup): ColumnGroup[] {
```

Comment:
Well, `separatePinnedColumnGroups` is a little inaccurate. Not all pinned groups will be split -- only those which have some pinned and some unpinned children. 

That's kind of the idea of the current name.. we're checking each group to see what needs to be done. Pin? Split for pinning? Nothing? It's different for each group depending on the state of the group's children.

Also, there is a helper method right above this one that does the actual splitting, and it's called `splitColumnGroupForPinning`.